src/main/java/mate/project/service/UserServiceImpl.java

```diff
@@ -0,0 +1,29 @@
+package mate.project.service;
+
+import lombok.RequiredArgsConstructor;
+import mate.project.dto.UserRegistrationRequestDto;
+import mate.project.dto.UserResponseDto;
+import mate.project.exception.RegistrationException;
+import mate.project.mapper.UserMapper;
+import mate.project.model.User;
+import mate.project.repository.user.UserRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()) != null) {
```

Comment:
The check for an existing user can lead to a race condition where two users with the same email can be created at the same time. Consider improving the uniqueness check at the database level, not at the service level.

Reply:
I thought the unique = true annotation for email in User model class also makes sure that the email is unique at the database level throwing the DataIntegrityViolationException which I now have also included to be handled in UserService hope this is a good solution